presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java

```diff
@@ -0,0 +1,284 @@
+
+public class InequalityInference
+{
+    // inequalityExpressions include the inequalities from the current join predicate
+    // and those inherited from the join's parent
+    private final Set<RowExpression> inequalityExpressions;
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final ExpressionEquivalence expressionEquivalence;
+    // 'outerVariables' is the variable set projected from the outer input for left or right join,
+    // or EMPTY if the join is an inner join
+    private final Optional<Collection<VariableReferenceExpression>> outerVariables;
+
+    public InequalityInference(Set<RowExpression> inequalityExpressions, FunctionAndTypeManager functionAndTypeManager, ExpressionEquivalence expressionEquivalence, Optional<Collection<VariableReferenceExpression>> outerVariables)
+    {
+        if (inequalityExpressions.stream()
+                .anyMatch(e -> !isOperation(e, LESS_THAN, functionAndTypeManager) && !isOperation(e, LESS_THAN_OR_EQUAL, functionAndTypeManager))) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "all inequality expressions for inference must be < or <=");
+        }
+        this.inequalityExpressions = requireNonNull(inequalityExpressions, "inequalityExpressions is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.expressionEquivalence = requireNonNull(expressionEquivalence, "expressionEquivalence is null");
+        this.outerVariables = requireNonNull(outerVariables, "outerVariables in null");
+    }
+
+    // Basic idea is to pairwise compare each predicate with all other predicates.
+    // Since any inferred predicates may trigger other inferences, keep comparing until no new predicates are generated
+    public Set<RowExpression> inferInequalities()
+    {
+        if (inequalityExpressions.size() < 2) {
+            return ImmutableSet.of();
+        }
+
+        Set<RowExpression> allInferredInequalities = new HashSet<>();
+        Set<RowExpression> inequalitiesInferredInCurrentTraversal = new HashSet<>(inequalityExpressions);
+        Set<Set<RowExpression>> exploredCombinations = new HashSet();
+
+        while (!allInferredInequalities.containsAll(inequalitiesInferredInCurrentTraversal)) {
+            allInferredInequalities.addAll(inequalitiesInferredInCurrentTraversal);
+            Set<Set<RowExpression>> newCombinations = Sets.combinations(allInferredInequalities, 2).stream()
+                    .filter(subset -> !exploredCombinations.contains(subset))
+                    .collect(Collectors.toSet());
+
+            for (Set<RowExpression> pair : newCombinations) {
+                Iterator<RowExpression> it = pair.iterator();
+                compareAndExtractInequalities(it.next(), it.next()).ifPresent(inequalitiesInferredInCurrentTraversal::add);
+            }
+
+            inequalitiesInferredInCurrentTraversal = newCombinations.stream()
```

Comment:
I think the code starting from line 96 and the code starting from here are trying to do the same thing, and we need only one or the other.